src/Allegro.Extensions.DependencyCall/Allegro.Extensions.DependencyCall/DependencyCall.cs

```
@@ -0,0 +1,98 @@
+
+internal interface IDependencyCall<in TRequest, TResponse> where TRequest : IRequest<TResponse>
+{
+    Task<TResponse> Run(
+        TRequest request,
+        IDependencyCallMetrics dependencyCallMetrics,
+        CancellationToken? cancellationToken = null);
+}
+
+#pragma warning disable MA0049
+/// <summary>
+/// Abstraction to support any dependency call, that allows to declare some common aspects (ex. fallbacks, metrics, retries, timeout) of any dependency call.
+/// </summary>
+/// <typeparam name="TRequest">Type of request data</typeparam>
+/// <typeparam name="TResponse">Type of response data</typeparam>
+public abstract class DependencyCall<TRequest, TResponse> : IDependencyCall<TRequest, TResponse>
+#pragma warning restore MA0049
+    where TRequest : IRequest<TResponse>
+{
+    private const int DefaultCancelCallAfterSeconds = 5;
+
+    async Task<TResponse> IDependencyCall<TRequest, TResponse>.Run(
+        TRequest request,
+        IDependencyCallMetrics dependencyCallMetrics,
+        CancellationToken? cancellationToken)
+    {
+        using var cancellationTokenSource = new CancellationTokenSource();
+        cancellationTokenSource.CancelAfter(CancelAfter);
+        cancellationToken ??= cancellationTokenSource.Token;
+
+        var policy = CustomPolicy(cancellationToken.Value);
+
+        using var dependencyCallTimer = dependencyCallMetrics.StartTimer(request);
+        try
+        {
+            dependencyCallMetrics.Requested(request);
+            var response = await policy.ExecuteAsync(token => Execute(request, token), cancellationToken.Value);
+            dependencyCallMetrics.Executed(request);
+            return response;
+        }
+        catch (Exception exception)
+        {
+            var fallbackValue = await Fallback(request, exception, cancellationToken.Value);
+            if (fallbackValue.ShouldThrowOnError == ShouldThrowOnError.Yes)
+            {
+                dependencyCallMetrics.Failed(request, exception);
```

`Failed` metric is not called when Fallback throws, because exceptions are not handled for Fallback.

---

you assume that fallback will be mostly calling other apis, I assume that some precofigured value should be returned, but yes, you are right

---

added tryFallback with wrapping both errors in one if error appears